libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -754,6 +754,40 @@ string YulUtilFunctions::storageArrayIndexAccessFunction(ArrayType const& _type)
 	});
 }
 
+string YulUtilFunctions::memoryArrayIndexAccessFunction(ArrayType const& _type)
+{
+	string functionName = "memory_array_index_access_" + _type.identifier();
+	return m_functionCollector->createFunction(functionName, [&]() {
+		return Whiskers(R"(
+			function <functionName>(baseRef, index) -> addr {
+				if iszero(lt(index, <arrayLen>(baseRef))) {
+					invalid()
+				}
+
+				let offset := index
+				<?noByteArray>
+					offset := mul(index, <headSize>)
+				</noByteArray>
+				<?dynamicallySized>
+					offset := add(offset, 32)
+				</dynamicallySized>
+				addr := add(baseRef, offset)
+			}
+		)")
+		("functionName", functionName)
+		("arrayLen", arrayLengthFunction(_type))
+		("headSize", to_string(_type.memoryHeadSize()))
+		("noByteArray", !_type.isByteArray())
+		("dynamicallySized", _type.isDynamicallySized())
+		.render();
+	});
+}
+
+string YulUtilFunctions::calldataArrayIndexAccessFunction(ArrayType const& /*_type*/)
+{
+	solUnimplemented("Calldata arrays not yet implemented!");
```

Comment:
The compiler doesn't complain that the end of the function is reached without return?

Reply:
Ah, ok, not even on my system. Great!